### ISSUE TITLE
No transparency when starting picom for the first time after boot

### ISSUE THREAD

**Issue by TRSx80** [yshui/picom#329]
First off, thanks for picking up the torch on the maintainership of this project! :beers:

Second, I am but a low level wizard, and very new to tiling WMs and etc., so I am probably doing something wrong in the way I am launching picom. But I will make my report anyhow.

### Platform

Debian Stable/Testing (happens on both) amd64

### GPU, drivers, and screen setup

Radeon R9 290, 3 monitors in a [20 30 20 PLP setup](https://www.startpage.com/do/dsearch?cat=pics&query=20%2030%2020%20PLP).

Output of `glxinfo -B`:
```
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: X.Org (0x1002)
    Device: AMD HAWAII (DRM 2.50.0, 5.4.0-4-amd64, LLVM 9.0.1) (0x67b1)
    Version: 19.3.3
    Accelerated: yes
    Video memory: 4096MB
    Unified memory: no
    Preferred profile: core (0x1)
    Max core profile version: 4.5
    Max compat profile version: 4.5
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.2
Memory info (GL_ATI_meminfo):
    VBO free memory - total: 3454 MB, largest block: 3454 MB
    VBO free aux. memory - total: 2030 MB, largest block: 2030 MB
    Texture free memory - total: 3454 MB, largest block: 3454 MB
    Texture free aux. memory - total: 2030 MB, largest block: 2030 MB
    Renderbuffer free memory - total: 3454 MB, largest block: 3454 MB
    Renderbuffer free aux. memory - total: 2030 MB, largest block: 2030 MB
Memory info (GL_NVX_gpu_memory_info):
    Dedicated video memory: 4096 MB
    Total available memory: 6139 MB
    Currently available dedicated video memory: 3454 MB
OpenGL vendor string: X.Org
OpenGL renderer string: AMD HAWAII (DRM 2.50.0, 5.4.0-4-amd64, LLVM 9.0.1)
OpenGL core profile version string: 4.5 (Core Profile) Mesa 19.3.3
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile

OpenGL version string: 4.5 (Compatibility Profile) Mesa 19.3.3
OpenGL shading language version string: 4.50
OpenGL context flags: (none)
OpenGL profile mask: compatibility profile

OpenGL ES profile version string: OpenGL ES 3.2 Mesa 19.3.3
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.20
```

I just spent about 30 minutes trying to figure out how to find "the version of the video drivers (xf86-video-*) and mesa" without success. But here is (what I thought were relevant) results of following command anyway(?):

```
$ sudo find / -iname *xf86*

/var/cache/apt/archives/x11proto-xf86vidmode-dev_2.3.1-2_all.deb
/var/cache/apt/archives/libxxf86vm-dev_1%3a1.1.4-1+b2_amd64.deb
```

If you give me some hints what to look for, I can try and get better information if you feel that's really relevant.

### Environment

This system was originally installed as Debian Stable with XFCE. Then I installed `i3-gaps` (compiled from latest source). Starting from lightdm greeter, launching desktop session goes like this:
* launch a helper script which:
  * sets wallpaper with `feh`
  * launches `xfsettingsd` to take care of:
	* dark theme on apps that need it
	* set screen orientations
	* launch dbus
	* and probably a bunch of other crap that I still do not fully understand (but was all set up in XFCE previously)
  * launches i3 via `dbus-launch i3-with-shmlog`
  * i3 config file loads, which in turn launches a bash `startup` script which:
	* launches picom via `dbus-launch picom -b --backend glx --experimental-backends`
	* moves around workspaces (using i3-msg) and launching several clients (terminals, Firefox, Emacs, etc.) in appropriate places

### picom version

Recent version, compiled from source IIRC.

This is interesting. Could be the whole problem right here.

```
2020-03-08 21:31 trs-80@hal:~
$ picom --version

vgit-a8e8b

2020-03-08 21:55 trs-80@hal:~
$ picom --diagnostics

[ 03/08/2020 21:55:16.047 get_cfg WARN ] --max-brightness requires the experimental glx backend. Falling back to 1.0
[ 03/08/2020 21:55:16.053 register_cm FATAL ERROR ] Another composite manager is already running

```

### Configuration:

You really want the whole thing? OK...

```
# Shadow
shadow = true;
shadow-radius = 7;
shadow-offset-x = -7;
shadow-offset-y = -7;
log-level = "warn";
# log-file = "/path/to/your/log/file";
# shadow-opacity = 0.7;
# shadow-red = 0.0;
# shadow-green = 0.0;
# shadow-blue = 0.0;
shadow-exclude = [
	"name = 'Notification'",
	"class_g = 'Conky'",
	"class_g ?= 'Notify-osd'",
	"class_g = 'Cairo-clock'",
	"_GTK_FRAME_EXTENTS@:c"
];
# shadow-exclude = "n:e:Notification";
# shadow-exclude-reg = "x10+0+0";
# xinerama-shadow-crop = true;

# Opacity
inactive-opacity = 0.85;
active-opacity = 0.95;
frame-opacity = 0.7;
inactive-opacity-override = false;
# inactive-dim = 0.2;
# inactive-dim-fixed = true;
# blur-background = true;
# blur-background-frame = true;
blur-kern = "3x3box";
# blur-kern = "5,5,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1";
# blur-background-fixed = true;
blur-background-exclude = [
	"window_type = 'dock'",
	"window_type = 'desktop'",
	"_GTK_FRAME_EXTENTS@:c"
];
opacity-rule = [
	#"100:class_g = 'Firefox-esr'",
	"100:class_g = 'gzdoom'",
	"80:class_g  = 'URxvt'"
];

# max-brightness = 0.66

# Fading
fading = true;
# fade-delta = 30;
fade-in-step = 0.03;
fade-out-step = 0.03;
# no-fading-openclose = true;
# no-fading-destroyed-argb = true;
fade-exclude = [ ];

# Other
backend = "xrender";
mark-wmwin-focused = true;
mark-ovredir-focused = true;
# use-ewmh-active-win = true;
detect-rounded-corners = true;
detect-client-opacity = true;
refresh-rate = 0;
vsync = true;
# sw-opti = true;
# unredir-if-possible = true;
# unredir-if-possible-delay = 5000;
# unredir-if-possible-exclude = [ ];
focus-exclude = [ "class_g = 'Cairo-clock'" ];
detect-transient = true;
detect-client-leader = true;
invert-color-include = [ ];
# resize-damage = 1;

# GLX backend
# glx-no-stencil = true;
# glx-no-rebind-pixmap = true;
# xrender-sync-fence = true;
use-damage = false;

# I don't think this does anything
max-brightness = 0.5;

# Window type settings
wintypes:
{
  tooltip = { fade = true; shadow = true; opacity = 1.0; focus = true; full-shadow = false; };
  dock = { shadow = false; }
  dnd = { shadow = false; }
  popup_menu = { opacity = 0.9; }
  dropdown_menu = { opacity = 0.9; }
};
```

### Steps of reproduction

1. Log in from lightdm.

### Expected behavior

Picom should launch on every startup / login.

### Current Behavior

Picom doesn't start after a cold boot (first login). But if I log out then back in, it does(?).

**Comment by yshui** [comment]
do you have any log files from the failed start?

**Comment by TRSx80** [comment]
This is reliably repeatable for me on a cold boot, so if you give me a hint as to how to generate logs (or where to find them), I am sure I could get you some.

**Comment by TRSx80** [comment]
https://paste.debian.net/hidden/81950f69/

OK, well, if I am reading this log correctly I guess `picom` *is* actually starting(?). I can't make much sense of it, maybe you can.

But none of the windows that the startup script open have any transparency. I also put `sleep 2` in the startup script after `picom` to try and give it more time to load, but that didn't seem to help. Maybe it has something to do with being called from the same script / process?

**Comment by yshui** [comment]
Yes, picom was started and running. If you have problem getting transparency, it's more likely a problem with your window manager. Can you maybe report this to i3-gaps and see what they say?

**Comment by TRSx80** [comment]
I haven't gotten around to reporting to i3-gaps yet, hopefully soon.

But I just wanted to report back that in the case where transparency doesn't work (what I opened this ticket for) the logging stops at ~ 80-90 kB and never gets any bigger.

Then if I exit and log back in through lightdm (and transparency works) I just noticed today that the log file continues to grow and grow. Looking at it in further detail, It seems to "start over" with similar logs as are seen in the beginning, and then eventually you start seeing some actual window names in the logs (Firefox, etc.). Where in the failure case, it never gets this far and the logging just stops.

I am not sure if that means picom is hanging or not. And then if so is that a picom problem or perhaps something with i3(-gaps)?

I posted some more logs. I had to snip out some from the middle because Debian paste limit is 150kB. You can see where I put `[SNIP]` but what I cut out was essentially redundant information, the same as in the first paste I linked above.

https://paste.debian.net/hidden/91391147/

Nome notes / thoughts (in parens):

```
[ 03/11/2020 10:08:15.583 atom_getter DEBUG ] (Start)
(same as before)
[SNIP]
(a little more stuff, same as before)
[ 03/11/2020 10:08:16.239 handle_pending_updates DEBUG ] (end of previous info)
(3 minute gap - presumably me logging out then back in)
[ 03/11/2020 10:11:43.714 atom_getter DEBUG ] (start of new info)
(eventually you see actual window names like Firefox, Terminal, emacs, etc.)
(I cut it off here, but log keeps going...)
```

One thing I noticed, upon second start of picom, this line:
```
[ 03/11/2020 10:11:43.714 atom_getter DEBUG ] Atom _NET_WM_WINDOW_OPACITY is 450
```
is different from the first start at:
```
[ 03/11/2020 10:08:15.583 atom_getter DEBUG ] Atom _NET_WM_WINDOW_OPACITY is 433
```
I have no ides whether this is significant or not?

**Comment by mighty9245** [comment]
> I just spent about 30 minutes trying to figure out how to find "the version of the video drivers (xf86-video-*) and mesa" without success. But here is (what I thought were relevant) results of following command anyway(?):

newbie :)
this is a task for you package manager (as i can see it's apt/dpkg and you're running debian). i didn't used apt-based distros for a while, but iirc one of ways to do this is `dpkg -l <query>`: `dpkg -l xf86` will return all packages matching xf86 with their versions and `dpkg -l mesa` is for mesa

you can make sure that picom is starting by configuring noticable visuals like big dark shadow, full transparency or something like that, i would recommend to make noticable shadows.

**Comment by TRSx80** [comment]
@mighty9245,

Thanks for the tip!

```
2020-03-11 11:51 user@host:~
$ dpkg-query -l *xf86*

Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                     Version      Architecture Description
+++-========================-============-============-============================================
ii  libxxf86dga1:amd64       2:1.1.4-1+b3 amd64        X11 Direct Graphics Access extension library
ii  libxxf86vm1:amd64        1:1.1.4-1+b2 amd64        X11 XFree86 video mode extension library
ii  libxxf86vm1:i386         1:1.1.4-1+b2 i386         X11 XFree86 video mode extension library
un  x11proto-xf86bigfont-dev <none>       <none>       (no description available)
un  x11proto-xf86dga-dev     <none>       <none>       (no description available)
un  x11proto-xf86dri-dev     <none>       <none>       (no description available)
un  x11proto-xf86vidmode-dev <none>       <none>       (no description available)

2020-03-11 11:51 user@host:~
$ dpkg-query -l *mesa*

Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                       Version      Architecture Description
+++-==========================-============-============-=========================================================>
ii  libegl-mesa0:amd64         19.3.3-1     amd64        free implementation of the EGL API -- Mesa vendor library
ii  libegl1-mesa:amd64         19.3.3-1     amd64        transitional dummy package
un  libegl1-mesa-dev           <none>       <none>       (no description available)
ii  libgl1-mesa-dev:amd64      19.3.3-1     amd64        transitional dummy package
ii  libgl1-mesa-dri:amd64      19.3.3-1     amd64        free implementation of the OpenGL API -- DRI modules
ii  libgl1-mesa-dri:i386       19.3.3-1     i386         free implementation of the OpenGL API -- DRI modules
ii  libgl1-mesa-glx:amd64      19.3.3-1     amd64        transitional dummy package
un  libgl1-mesa-swx11          <none>       <none>       (no description available)
ii  libglapi-mesa:amd64        19.3.3-1     amd64        free implementation of the GL API -- shared library
ii  libglapi-mesa:i386         19.3.3-1     i386         free implementation of the GL API -- shared library
un  libgles2-mesa-dev          <none>       <none>       (no description available)
ii  libglu1-mesa:amd64         9.0.1-1      amd64        Mesa OpenGL utility library (GLU)
ii  libglu1-mesa:i386          9.0.1-1      i386         Mesa OpenGL utility library (GLU)
ii  libglu1-mesa-dev:amd64     9.0.1-1      amd64        Mesa OpenGL utility library -- development files
ii  libglx-mesa0:amd64         19.3.3-1     amd64        free implementation of the OpenGL API -- GLX vendor libra>
ii  libglx-mesa0:i386          19.3.3-1     i386         free implementation of the OpenGL API -- GLX vendor libra>
ii  libosmesa6:amd64           19.3.3-1     amd64        Mesa Off-screen rendering extension
ii  libosmesa6:i386            19.3.3-1     i386         Mesa Off-screen rendering extension
ii  libwayland-egl1-mesa:amd64 19.3.3-1     amd64        transitional dummy package
un  mesa-common-dev            <none>       <none>       (no description available)
un  mesa-glide2-dev            <none>       <none>       (no description available)
un  mesa-opencl-icd            <none>       <none>       (no description available)
ii  mesa-utils                 8.4.0-1+b1   amd64        Miscellaneous Mesa GL utilities
ii  mesa-va-drivers:amd64      19.3.3-1     amd64        Mesa VA-API video acceleration drivers
ii  mesa-va-drivers:i386       19.3.3-1     i386         Mesa VA-API video acceleration drivers
ii  mesa-vdpau-drivers:amd64   19.3.3-1     amd64        Mesa VDPAU video acceleration drivers
ii  mesa-vdpau-drivers:i386    19.3.3-1     i386         Mesa VDPAU video acceleration drivers
ii  mesa-vulkan-drivers:amd64  19.3.3-1     amd64        Mesa Vulkan graphics drivers
ii  mesa-vulkan-drivers:i386   19.3.3-1     i386         Mesa Vulkan graphics drivers
un  mesag-dev                  <none>       <none>       (no description available)
un  mesag3                     <none>       <none>       (no description available)
un  mesag3+ggi-dev             <none>       <none>       (no description available)
un  xlibmesa-dev               <none>       <none>       (no description available)
un  xlibmesa-glu-dev           <none>       <none>       (no description available)
un  xlibmesa3                  <none>       <none>       (no description available)

2020-03-11 11:52 user@host:~
$ 
```

**Comment by yshui** [comment]
@TRSx80 hmm, thanks for the additional information.

In the not working case, can you maybe attach `gdb` to picom and see where it's at? maybe `strace` too.

**Comment by TRSx80** [comment]
OK, so the other day I installed `strace` and got some more logs. I start `picom` via `dbus-launch` and so I ran `strace` both before and after the `dbus-launch` as I wasn't sure what would be relevant.

If I understand correctly, whatever command you put immediately following `strace` is the one that gets traced. So essentially, the first command I did below should trace `dbus-launch` and the second should trace actual `picom`.

I turned on microsecond times (`-tt`) in `strace` logs in hopes it might be helpful correlating something with `picom` logs.

Here are the commands I ran and the resulting logs:

## strace before dbus-launch
command: `strace -tt -o /home/user/tmp/picom-logs/strace-log-before-dbus dbus-launch picom -b --backend glx --experimental-backends --log-level debug --log-file /home/user/tmp/picom-logs/picom-log-strace-before-dbus`

associated strace log: https://termbin.com/7gf8

associated picom log: https://termbin.com/8pur

## strace after dbus-launch
command: `dbus-launch strace -tt -o /home/user/tmp/picom-logs/strace-log-after-dbus picom -b --backend glx --experimental-backends --log-level debug --log-file /home/user/tmp/picom-logs/picom-log-strace-after-dbus`

associated strace log: https://termbin.com/v8b9

associated picom log: https://termbin.com/mkf4

## Some more notes

I haven't tried to get anything from `gdb` yet. All of these tools are new to me, so maybe I work on that when I get some more time. In the meantime, please let me know if the `strace` logs are sufficient by themselves to learn anything useful. Thanks!

**Comment by TRSx80** [comment]
I am very pleased to report that I found the problem. There was some sort of conflict with `xfsettingsd` (the xfce x settings manager daemon) which I was using to handle setting GTK dark theme and some other things, for those applications that need it.

I needed a break from stack traces and trying (unsuccessfully thus far) to get `gdb` to work non-interactively, so I was playing around with `rofi` this evening, and noticed a weird issue with the workspace part of the listing in the window switcher. Some searching led me to another person having the [exact same issue](https://github.com/davatorium/rofi/issues/901). He was using `xfsettingsd` just like I was, and said when he disabled it those problems went away. So I did that too, and my `rofi` menu problem went away. On a hunch, I put back my regular settings (removed `gdb` stuff I was working on) and rebooted my computer, and now `picom` works just as it should from a cold boot, also.

I ended up coming across [`xsettingsd`](https://github.com/derat/xsettingsd) (note the slight spelling difference (no " F ") from above) which seems to be working as a minimal alternative to store settings for those few programs that need it, and was also available as a Debian package. :+1:

OK, now it's Miller Time... :beers: :smile: 

Closing... Thanks!
